master/internal/api_notebook.go

```diff
@@ -89,48 +96,120 @@ func (a *apiServer) GetNotebook(
 		return nil, err
 	}
 
-	if ok, err := user.AuthZProvider.Get().CanAccessNTSCTask(
-		ctx, *curUser, model.UserID(resp.Notebook.UserId)); err != nil {
+	if ok, err := command.AuthZProvider.Get().CanGetNSC(
+		ctx, *curUser, model.AccessScopeID(resp.Notebook.WorkspaceId),
+	); err != nil {
 		return nil, err
-	} else if !ok {
+	} else if !ok { // permission denied.
+		// report the error as if the notebook does not exist.
 		return nil, errActorNotFound(addr)
 	}
 	return resp, nil
 }
 
+func (a *apiServer) validateToKillNotebook(ctx context.Context, notebookID string) error {
+	targetNotebook, err := a.GetNotebook(ctx, &apiv1.GetNotebookRequest{NotebookId: notebookID})
+	if err != nil {
+		return err
+	}
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = command.AuthZProvider.Get().CanTerminateNSC(
+		ctx, *curUser, model.AccessScopeID(targetNotebook.Notebook.WorkspaceId),
+	)
+	return apiutils.MapAndFilterErrors(err, nil, nil)
+}
+
 func (a *apiServer) IdleNotebook(
 	ctx context.Context, req *apiv1.IdleNotebookRequest,
 ) (resp *apiv1.IdleNotebookResponse, err error) {
-	if _, err := a.GetNotebook(ctx,
-		&apiv1.GetNotebookRequest{NotebookId: req.NotebookId}); err != nil {
+	err = a.validateToKillNotebook(ctx, req.NotebookId)
+	if err != nil {
 		return nil, err
 	}
-
 	return resp, a.ask(notebooksAddr.Child(req.NotebookId), req, &resp)
 }
 
 func (a *apiServer) KillNotebook(
 	ctx context.Context, req *apiv1.KillNotebookRequest,
 ) (resp *apiv1.KillNotebookResponse, err error) {
-	if _, err := a.GetNotebook(ctx,
-		&apiv1.GetNotebookRequest{NotebookId: req.NotebookId}); err != nil {
+	err = a.validateToKillNotebook(ctx, req.NotebookId)
+	if err != nil {
 		return nil, err
 	}
-
 	return resp, a.ask(notebooksAddr.Child(req.NotebookId), req, &resp)
 }
 
 func (a *apiServer) SetNotebookPriority(
 	ctx context.Context, req *apiv1.SetNotebookPriorityRequest,
 ) (resp *apiv1.SetNotebookPriorityResponse, err error) {
-	if _, err := a.GetNotebook(ctx,
-		&apiv1.GetNotebookRequest{NotebookId: req.NotebookId}); err != nil {
+	targetNotebook, err := a.GetNotebook(ctx, &apiv1.GetNotebookRequest{NotebookId: req.NotebookId})
+	if err != nil {
 		return nil, err
 	}
 
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	err = command.AuthZProvider.Get().CanSetNSCsPriority(
+		ctx, *curUser, model.AccessScopeID(targetNotebook.Notebook.WorkspaceId), int(req.Priority),
+	)
+	if err != nil {
+		return nil, apiutils.MapAndFilterErrors(err, nil, nil)
```

Comment:
non-blocking: I like the defer style but this looks like it works too. If you want to change it, I think we can do it later.

Reply:
👍🏼  let's freeze this and add any non essential changes to https://github.com/determined-ai/determined/tree/rbac-ntsc instead